examples/connectivity/plot_ica_neurovault.py

```diff
@@ -0,0 +1,244 @@
+from nilearn.plotting import plot_stat_map
+
+
+# Use for caching results to disk, so running twice will be *fast*
+mem = Memory('cache', verbose=0)
+
+
+@mem.cache
+def splitext_multi(p):
+    """From my_img.nii.gz, return .nii.gz"""
+    dirname, basename = os.path.dirname(p), os.path.basename(p)
+    stem, ext = basename.split('.', 1)
+    return os.path.join(dirname, stem), '.' + ext if ext else ''
+
+
+@mem.cache
+def resample_and_expand_image(src_img, target_img, out_path=None):
+    """Resample src_img to target_img affine, turn 4D into list of 3D.
+
+    Optionally save to out_path.
+    """
+    data = src_img.get_data().squeeze()
+    data[np.isnan(data)] = 0
+    assert np.abs(data).max() > 0
+
+    if out_path is not None:
+        # Create the proper output filenames (if 3D, just one;
+        #   if 4D then one per time point)
+        out_stem, out_ext = splitext_multi(out_path)
+        out_paths = [out_path] if len(data.shape) == 3 else \
+                    ['%s-%d%s' % (out_stem, ii, out_ext)
+                     for ii in range(data.shape[-1])]
+        # If all output images exist, use them cached from disk.
+        if np.all([os.path.exists(p) for p in out_paths]):
+            return [nib.load(p) for p in out_paths]
+
+    # Copy the image and make a background mask.
+    src_img = nib.Nifti1Image(data, src_img.get_affine(),
+                              header=src_img.get_header())
+    bg_mask = compute_background_mask(src_img).get_data()
+
+    # Test if the image has been masked:
+    out_of_mask = data[np.logical_not(bg_mask)]
+    if np.all(np.isnan(out_of_mask)) or len(np.unique(out_of_mask)) == 1:
+        # Need to extrapolate
+        with warnings.catch_warnings(record=False):
+            warnings.simplefilter('ignore', RuntimeWarning)
+            data, _ = _extrapolate_out_mask(data.astype(np.float), bg_mask,
+                                            iterations=3)
+    src_img = nib.Nifti1Image(data, src_img.get_affine(),
+                              header=src_img.get_header())
+    del out_of_mask, bg_mask
+
+    # Resampling the file to target and saving the output in the "resampled"
+    # folder
+    resampled_nii = resample_img(src_img, target_img.get_affine(),
+                                 target_img.shape)
+    resampled_nii = nib.Nifti1Image(resampled_nii.get_data().squeeze(),
+                                    resampled_nii.get_affine(),
+                                    header=src_img.get_header())
+
+    # Decimate 4D files.
+    if len(resampled_nii.shape) == 3:
+        resampled_niis = [resampled_nii]
+    else:
+        resampled_niis = []
+        resampled_data = resampled_nii.get_data()
+        for index in range(resampled_nii.shape[3]):
+            # First save the files separately
+            this_nii = nib.Nifti1Image(resampled_data[..., index],
+                                       resampled_nii.get_affine())
+            resampled_niis.append(this_nii)
```

Comment:
One downside of `iter_img` is that it fails on a 3D image, rather than giving me an iterator for the one image. So I'll have to load the image and check if it's 4D before calling `iter_img`... that seems clunky to me.